src/sparse_blas/backends/cusparse/cusparse_handles.cpp

```diff
@@ -0,0 +1,520 @@
+}
+
+template <typename fpType, typename intType>
+void set_coo_matrix_data(sycl::queue &queue, matrix_handle_t smhandle, std::int64_t num_rows,
+                         std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index,
+                         sycl::buffer<intType, 1> row_ind, sycl::buffer<intType, 1> col_ind,
+                         sycl::buffer<fpType, 1> val) {
+    detail::check_can_reset_sparse_handle<fpType, intType>(__func__, smhandle, true);
+    auto event = queue.submit([&](sycl::handler &cgh) {
+        auto row_acc = row_ind.template get_access<sycl::access::mode::read_write>(cgh);
+        auto col_acc = col_ind.template get_access<sycl::access::mode::read_write>(cgh);
+        auto val_acc = val.template get_access<sycl::access::mode::read_write>(cgh);
+        submit_host_task(cgh, queue, [=](CusparseScopedContextHandler &sc) {
+            // Ensure that a cusparse handle is created before any other cuSPARSE function is called.
+            sc.get_handle(queue);
+            if (smhandle->num_rows != num_rows || smhandle->num_cols != num_cols ||
+                smhandle->nnz != nnz || smhandle->index != index) {
+                CUSPARSE_ERR_FUNC(cusparseDestroySpMat, smhandle->backend_handle);
+                auto cuda_index_type = CudaIndexEnumType<intType>::value;
+                auto cuda_index_base = get_cuda_index_base(index);
+                auto cuda_value_type = CudaEnumType<fpType>::value;
+                CUSPARSE_ERR_FUNC(cusparseCreateCoo, &smhandle->backend_handle, num_rows, num_cols,
+                                  nnz, sc.get_mem(row_acc), sc.get_mem(col_acc),
+                                  sc.get_mem(val_acc), cuda_index_type, cuda_index_base,
+                                  cuda_value_type);
+                smhandle->num_rows = num_rows;
+                smhandle->num_cols = num_cols;
+                smhandle->nnz = nnz;
+                smhandle->index = index;
+            }
+            else {
+                CUSPARSE_ERR_FUNC(cusparseCooSetPointers, smhandle->backend_handle,
+                                  sc.get_mem(row_acc), sc.get_mem(col_acc), sc.get_mem(val_acc));
+            }
+            smhandle->row_container.set_buffer(row_ind);
+            smhandle->col_container.set_buffer(col_ind);
+            smhandle->value_container.set_buffer(val);
+        });
+    });
+    event.wait_and_throw();
+}
+
+template <typename fpType, typename intType>
+void set_coo_matrix_data(sycl::queue &queue, matrix_handle_t smhandle, std::int64_t num_rows,
+                         std::int64_t num_cols, std::int64_t nnz, oneapi::mkl::index_base index,
+                         intType *row_ind, intType *col_ind, fpType *val) {
+    detail::check_can_reset_sparse_handle<fpType, intType>(__func__, smhandle, false);
+    auto event = queue.submit([&](sycl::handler &cgh) {
+        submit_host_task(cgh, queue, [=](CusparseScopedContextHandler &sc) {
```

Comment:
These are all functions that don't take in `dependencies` and don't return a `sycl::event`, but internally submit a `host_task` and immediately wait on that event to complete before returning. Why does it have to go through the `host_task`? Does it have something to do with `CusparseScopedContextHandler`, or something that simply can't directly be done on the host without a `host_task`?

Reply:
I can't think of any **function** in the sparse BLAS domain that can work without an `sparse::init_*_matrix()`/`sparse::init_*_vector()`/`sparse::init_<api>_descr()` API being called first to create either a matrix handle or an API descriptor object. (Of course, variables like `enum`s can be declared, but I doubt that would break cuSPARSE).

Any non-`init*` sparse BLAS API being called without such an `init*` API call first would either segfault or crash because the data structure variable (think: `spMat` in `gemv(queue, ..., spMat, ...)` or in `create_csr_matrix(queue, ..., spMat, ...)`) would either be a `nullptr` or undefined garbage value.

That's why I feel your calls in `init*` APIs can keep doing what you have implemented, but would additionally store a pointer to the `cusparesHandle_t` object you are creating in the data structure being initialized.

I'm not saying we _should_ go down this route, but _if_ you think that would simplify things greatly on how events are submitted, then maybe its worth exploring

Reply:
Resolving as per https://github.com/oneapi-src/oneMKL/pull/527#discussion_r1754748497